File: sdks/typescript/tests/signing.test.ts
import { describe, it, expect } from "vitest";
import { signBody, signPayload, publicKeyBase64, publicKeyHex } from "../src/signing.js";

const SEED = new TextEncoder().encode("test-seed-32-bytes-for-testing!!");

describe("signing", () => {
  it("signBody returns base64 of 64-byte signature", async () => {
    const sig = await signBody(SEED, new TextEncoder().encode('{"hello":"world"}'));
    const raw = Buffer.from(sig, "base64");
    expect(raw.length).toBe(64);
  });

  it("signPayload is deterministic", async () => {
    const payload = { gene_id: "g1", confidence: 0.9 };
    const sig1 = await signPayload(SEED, payload);
    const sig2 = await signPayload(SEED, payload);
    expect(sig1).toBe(sig2);
    expect(Buffer.from(sig1, "base64").length).toBe(64);
  });

  it("publicKeyBase64 returns 32 bytes encoded", async () => {
    const pk = await publicKeyBase64(SEED);
    const raw = Buffer.from(pk, "base64");
    expect(raw.length).toBe(32);
  });

  it("publicKeyHex returns 64 hex chars", async () => {
    const pk = await publicKeyHex(SEED);
    expect(pk.length).toBe(64);
    expect(/^[0-9a-f]+$/.test(pk)).toBe(true);
  });
});

File: sdks/java/src/test/java/io/oris/sdk/SigningTest.java
package io.oris.sdk;

import io.oris.sdk.signing.Ed25519Signing;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class SigningTest {

    private static final byte[] TEST_SEED = new byte[32];

    static {
        for (int i = 0; i < 32; i++) TEST_SEED[i] = (byte) (i + 1);
    }

    @Test
    void signBodyProducesBase64Signature() {
        byte[] body = "hello world".getBytes();
        String sig = Ed25519Signing.signBody(TEST_SEED, body);
        assertNotNull(sig);
        assertFalse(sig.isEmpty());
        // Ed25519 signature is 64 bytes = 88 chars base64
        assertEquals(88, sig.length());
    }

    @Test
    void signPayloadSerializesAndSigns() {
        Map<String, Object> payload = Map.of("gene_id", "g1", "confidence", 0.9);
        String sig = Ed25519Signing.signPayload(TEST_SEED, payload);
        assertNotNull(sig);
        assertEquals(88, sig.length());
    }

    @Test
    void publicKeyBase64Returns44Chars() {
        String pk = Ed25519Signing.publicKeyBase64(TEST_SEED);
        assertNotNull(pk);
        assertEquals(44, pk.length());
    }

    @Test
    void publicKeyHexReturns64Chars() {
        String pk = Ed25519Signing.publicKeyHex(TEST_SEED);
        assertNotNull(pk);
        assertEquals(64, pk.length());
        assertTrue(pk.matches("[0-9a-f]+"));
    }

    @Test
    void deterministicSignatures() {
        byte[] body = "deterministic test".getBytes();
        String sig1 = Ed25519Signing.signBody(TEST_SEED, body);
        String sig2 = Ed25519Signing.signBody(TEST_SEED, body);
        assertEquals(sig1, sig2);
    }
}
